src/coreclr/src/tools/Common/Sorting/MergeSortCore.cs

```diff
@@ -0,0 +1,130 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System;
+using System.Collections.Generic;
+using System.Threading.Tasks;
+
+namespace ILCompiler.Sorting.Implementation
+{
+    internal static class MergeSortCore<T, TDataStructure, TDataStructureAccessor, TComparer, TCompareAsEqualAction>
+        where TDataStructureAccessor:ISortableDataStructureAccessor<T, TDataStructure>
+        where TComparer:IComparer<T>
+        where TCompareAsEqualAction : ICompareAsEqualAction
+    {
+        internal const int ParallelSortThreshold = 4000; // Number empirically measured by compiling
+                                                         // a large composite binary
+
+        public static void ParallelSortApi(TDataStructure arrayToSort, TComparer comparer)
+        {
+            TDataStructureAccessor accessor = default(TDataStructureAccessor);
+            if (accessor.GetLength(arrayToSort) < ParallelSortThreshold)
+            {
+                // If the array is sufficiently small as to not need parallel sorting
+                // call the sequential algorithm directly, as the parallel api will create
+                // an unnecessary Task object to wait on
+                SequentialSort(arrayToSort, 0, accessor.GetLength(arrayToSort), comparer);
+            }
+            ParallelSort(arrayToSort, 0, accessor.GetLength(arrayToSort), comparer).Wait();
+        }
+
+        // Parallelized merge sort algorithm. Uses Task infrastructure to spread sort across available resources
+        private static async Task ParallelSort(TDataStructure arrayToSort, int index, int length, TComparer comparer)
+        {
+            if (length < ParallelSortThreshold)
+            {
+                SequentialSort(arrayToSort, index, length, comparer);
+            }
+            else
+            {
+                TDataStructureAccessor accessor = default(TDataStructureAccessor);
+                int halfLen = length / 2;
+
+                TaskCompletionSource<bool> rightSortComplete = new System.Threading.Tasks.TaskCompletionSource<bool>();
+                _ = Task.Run(async () =>
+                {
+                    await ParallelSort(arrayToSort, index + halfLen, length - halfLen, comparer);
+                    rightSortComplete.SetResult(true);
+                });
+
+                T[] localCopyOfHalfOfArray = new T[halfLen];
+                accessor.Copy(arrayToSort, index, localCopyOfHalfOfArray, 0, halfLen);
+                await MergeSortCore<T, T[], ArrayAccessor<T>, TComparer, TCompareAsEqualAction>.ParallelSort(localCopyOfHalfOfArray, 0, halfLen, comparer);
+                await rightSortComplete.Task;
```

Comment:
Why not save the task returned from `Task.Run` and await that here?

Reply:
The Task returned by `Task.Run` completes at the first await, and not when the SetResult completes, therefore it isn't useful to await that.

Reply:
The `Func<Task>` overload of `Task.Run` returns a proxy task that represents the full completion ([MSDN](https://docs.microsoft.com/en-us/dotnet/api/system.threading.tasks.task.run?view=netcore-3.1#System_Threading_Tasks_Task_Run__1_System_Func_System_Threading_Tasks_Task___0___)).